### DISCUSSION TITLE
Looking at Schema.org for inspiration

### DISCUSSION THREAD

**Discussion by bmann** [lexicon-community/lexicon#19] [General]
The Recipe Exchange [Lexicon](https://recipe.exchange/lexicons/) has this note:

> Our recipe lexicon is designed to be compatible with Schema.org's Recipe type, ensuring that recipes shared on recipe.exchange can be easily converted to and from Schema.org's widely-adopted format. This compatibility enables better interoperability with existing recipe platforms and search engines.
> 
> Key properties from Schema.org that we've incorporated include:
>
> * Recipe metadata (name, description, yield, category, cuisine)
> * Time-related properties (prepTime, cookTime, totalTime)
> * Nutritional information
> * Dietary restrictions
> * Ingredients and instructions

Perhaps a good idea to look at more [Schema.org](https://schema.org/) domains and primitives. https://schema.org/docs/schemas.html

In fact, we may want to introduce ourselves at some point.

> Schema.org vocabulary can be used with many different encodings, including RDFa, Microdata and JSON-LD. These vocabularies cover entities, relationships between entities and actions, and can easily be extended through a well-documented extension model. As of 2024, over 45 million web domains markup their web pages with over 450 billion Schema.org objects. Many applications from Google, Microsoft, Pinterest, Yandex and others already use these vocabularies to power rich, extensible experiences.

* https://www.w3.org/community/schemaorg/
* https://github.com/schemaorg/schemaorg

[Open Graph](https://ogp.me/) is the other one to look at.

**Comment by rektide** [comment]
Such a large vocabulary/lexicon, used across the vastness of the web & email!

I'd love to see an org.schema lexicon that unambiguously can claim to be a faithful lexicon rendition.

**Comment by darccio** [comment]
I have been investigating this idea these past weeks, and I think it would bring a lot of potential to ATProto ecosystem.

The only thing that I'm not completely sure if it should be a direct translation as is or a subset that picks relevant fields for specific schemas from there while keeping 1:1 compatibility.

**Reply by darccio** [reply]
I took a first stab to migrating Schema.org's `LocalBusiness`: https://github.com/atblueprints/lexicons/blob/main/at/blueprints/localbusiness.json 

I also took some time to draft a first CUE definition for ATProto lexicons, so it was quite a breeze to write the specific `LocalBusiness` lexicon. Anyway, the more I dig, the more I feel a 1:1 translation isn't the best option.

Edit: an example of what I suggest is that Schema.org doesn't have the concept of `category` for this type, but both [Google Business Profile](https://www.google.com/intl/en_uk/business/articles/how-to-add-your-business-to-google-maps/) and [Open Street Maps](https://wiki.openstreetmap.org/wiki/Map_features) expect you to fit your business in a category.

**Comment by ngerakines** [comment]
Schema.org has a lot to offer. The Smoke Signal lexicon borrows from it, specifically for event status and mode. However, I decided to cherry-pick elements only as needed. I didn't want to add too much too quickly, and I think schema.org has some of the same issues as ActivityStreams / ActivityStreams-Vocab.

**Comment by ngerakines** [comment]
From https://github.com/lexicon-community/lexicon/pull/20#issuecomment-2557956362 , I think the only thing that stands out in AT Protocol vs [ActivityStreams](https://www.w3.org/TR/activitystreams-core/#intransitiveactivities) is that in AT Protocol the field is usually called `subject` where in others it is named `object`.

Personally, I like `subject` because I think it more clearly states that the reference in the field is the subject of the intention / action.

**Comment by darobin** [comment]
So, this is probably very stupid but I was working on schema.org for another project (and have used it a lot before) so I wanted to see how far / how easily it could be automatically translated to Lexicon. Perfect tiny project to hack on as conference knitting. The result is https://github.com/darobin/schema.lexicon.

Translating the entire schema.org schema (including extensions) yields [an 18mb lexicon](https://github.com/darobin/schema.lexicon/blob/main/schema.lexicon.json) which is a bit much.

Some points of friction have been noted upthread. More basic ones are 1) basically everything in schema.org can be an array or a scalar and 2) it's pretty common to have unions between structured and scalar types, which in lexicon requires a workaround (using `@value`, which is icky).

Evidently, no translation is needed if the idea is to just use it for inspiration. Having a clear mapping become more interesting if there is potential to interoperate with schema.org, for instance with something like [NLWeb](https://nlweb.ai/) or (maybe) a schema.orgified [Beckn Protocol.](https://becknprotocol.io/)
